aptos-move/aptos-vm/src/move_vm_ext/vm.rs

```diff
@@ -1,175 +1,186 @@
+}
+
+impl GenesisMoveVM {
+    pub fn new(chain_id: ChainId) -> Self {
+        let features = Features::default();
+        let timed_features = TimedFeaturesBuilder::enable_all().build();
+
+        // Genesis runs sessions, where there is no concept of block execution.
+        // Hence, delayed fields are not enabled.
+        let delayed_field_optimization_enabled = false;
+        let vm_config = aptos_prod_vm_config(
+            &features,
+            &timed_features,
+            delayed_field_optimization_enabled,
+        );
+
+        // All genesis sessions run with unmetered gas meter, and here we set
+        // the gas parameters for natives as zeros (because they do not matter).
+        let mut native_builder = SafeNativeBuilder::new(
+            LATEST_GAS_FEATURE_VERSION,
+            NativeGasParameters::zeros(),
+            MiscGasParameters::zeros(),
+            timed_features.clone(),
+            features.clone(),
+        );
+
+        let vm = MoveVM::new_with_config(
+            aptos_natives_with_builder(&mut native_builder),
+            vm_config.clone(),
+        );
+
+        Self {
+            vm,
+            chain_id,
+            features,
+        }
+    }
+
+    pub fn genesis_change_set_configs(&self) -> ChangeSetConfigs {
+        // Because genesis sessions are not metered, there are no change set
+        // (storage) costs as well.
+        ChangeSetConfigs::unlimited_at_gas_feature_version(LATEST_GAS_FEATURE_VERSION)
+    }
+
+    pub fn new_genesis_session<'r, R: AptosMoveResolver>(
+        &self,
+        resolver: &'r R,
+        genesis_id: HashValue,
+    ) -> SessionExt<'r, '_> {
+        let session_id = SessionId::genesis(genesis_id);
+        SessionExt::new(
+            session_id,
+            &self.vm,
+            self.chain_id,
+            &self.features,
+            None,
+            resolver,
+        )
+    }
+}
 
 pub struct MoveVmExt {
     inner: MoveVM,
-    chain_id: u8,
-    features: Features,
+    pub(crate) env: Arc<Environment>,
 }
 
 impl MoveVmExt {
     fn new_impl<F>(
         native_gas_params: NativeGasParameters,
         misc_gas_params: MiscGasParameters,
         gas_feature_version: u64,
-        chain_id: u8,
-        features: Features,
-        timed_features: TimedFeatures,
+        env: Arc<Environment>,
         gas_hook: Option<F>,
         resolver: &impl AptosMoveResolver,
-        aggregator_v2_type_tagging: bool,
-    ) -> VMResult<Self>
+    ) -> Self
     where
         F: Fn(DynamicExpression) + Send + Sync + 'static,
     {
         let mut builder = SafeNativeBuilder::new(
             gas_feature_version,
             native_gas_params.clone(),
             misc_gas_params.clone(),
-            timed_features.clone(),
-            features.clone(),
+            env.timed_features.clone(),
+            env.features.clone(),
         );
         if let Some(hook) = gas_hook {
             builder.set_gas_hook(hook);
```

Comment:
Based on little grep-ing, this might be the only place set_gas_hook is used, in which case I don't quite understand why the setter method is needed at all, if it's only call after the SateNativeBuilder constructor which sets self.gas_hook to None and then overwrites it. Let's just pass gas_hook (or maybe_gas_hook better name) to the constructor and get rid of this setter.